File: examples/python/kmeans/plotting/bv_annz_evaluation.py
from __future__ import print_function
import fcl
import matplotlib
import os
import matplotlib.pyplot as plt
from fcl import kmeans
from fcl.datasets import load_sector_dataset, load_usps_dataset
from os.path import abspath, join, dirname

def do_evaluations(datasets):
  
  dataset_results = {}
  for dataset_name, dataset_as_string in datasets:
    print("Doing evaluations for dataset %s"%dataset_name)
    dataset_results[dataset_name] = {'searching_best_b': {'durations': [], 'bv_annz': []}}
    
    for i in range(1,101, 10):
      bv_annz = float(i)/100
      print("Executing k-means optimized with bv_annz: %f"%bv_annz)
      km = kmeans.KMeans(n_jobs=1, no_clusters=1000, algorithm="kmeans_optimized", init='random', additional_params = {'bv_annz': bv_annz}, seed = 0, verbose = False)
      km.fit(dataset_as_string)
      dataset_results[dataset_name]['searching_best_b']['durations'].append(km.get_tracked_params()['duration_kmeans'] / 1000)
      dataset_results[dataset_name]['searching_best_b']['bv_annz'].append(km.get_tracked_params()['additional_params']['bv_annz'])
  
  for dataset_name in dataset_results:
    plt.plot(dataset_results[dataset_name]['searching_best_b']['bv_annz'],
             dataset_results[dataset_name]['searching_best_b']['durations'], '-', linewidth=3, label = dataset_name)

  plt.legend()
  plt.grid(True)
  
  plt.xlabel('relative block vector size')
  plt.ylabel('time / s')
  plt.title(r'Varying the block vector size in algorithm kmeans-optimized')
  
  destination_filename = join(dirname( __file__ ), "bv_annz_evaluation.png")
  plt.savefig(destination_filename)
  
  print("plot was saved in the current folder to: %s"%destination_filename)

if __name__ == "__main__":
  datasets = []
  
  ds_folder = abspath(join(dirname( __file__ ), os.pardir, os.pardir, os.pardir, 'datasets'))
  
  with open(load_sector_dataset(ds_folder), 'r') as f:
    datasets.append(('sector', f.read()))
  
  with open(load_usps_dataset(ds_folder), 'r') as f:
    datasets.append(('usps', f.read()))

  do_evaluations(datasets)
  
  

File: examples/matlab/kmeans/plotting/bv_annz_evaluation.m
function bv_annz_evaluation()   
    addpath(fileparts(fileparts(fileparts(mfilename('fullpath')))));
    check_compile('kmeans')
    datasets.sector = get_dataset('sector.scaled');
    datasets.usps = get_dataset('usps.scaled');
    do_bv_annz_evaluation(datasets)
end

function do_bv_annz_evaluation(datasets)
    opts.silent = true;
    opts.no_cores = 1;
    opts.seed = 0;
    opts.algorithm = 'bv_kmeans';
    
    bv_annz_values = linspace(1, 100, 10) / 100.0;
    
    dataset_names = fieldnames(datasets);
    result_durations = cell(numel(dataset_names),1);
    result_bv_annz = cell(numel(dataset_names),1);
    for i = 1:numel(dataset_names)
      dataset_path = datasets.(dataset_names{i});
      for bv_annz = bv_annz_values
          additional_params.bv_annz = bv_annz;
          opts.additional_params = additional_params;
          fprintf('Executing %s for %s with bv_annz: %f\n', opts.algorithm, dataset_names{i}, bv_annz);
          [ C, tracked_params ] = fcl_kmeans_fit(dataset_path, 1000, opts);
          result_durations{i} = [result_durations{i} tracked_params.duration_kmeans];
          result_bv_annz{i} = [result_bv_annz{i} tracked_params.additional_params.bv_annz];
      end
    end
    
    figure('DefaultAxesFontSize',12,'Position', [100, 100, 800, 800]);
    cc=hsv(numel(dataset_names));
    hold on;
    for i = 1:numel(dataset_names)
        plot(result_bv_annz{i}, result_durations{i} / 1000.0, 'color', cc(i,:), 'LineWidth', 3);
    end
    hold off;
    legend(dataset_names, 'Interpreter', 'none');
    xlabel('relative block vector size');
    ylabel('time / s');
    title('Observing algorithm speed while varying block vector size');
end
